### ISSUE TITLE
Window to front

### ISSUE THREAD

**Issue by SirWindfield** [andlabs/ui#194]
Is there some way to bring the window to the front programmatically?

**Comment by andlabs** [comment]
What OS? What scenario?

**Comment by SirWindfield** [comment]
It happens on OS X to me, I run the created executable from the command line. The window opens and centers as expected, but the focus doesnt switch to the application. The active application is still the terminal.

**Comment by andlabs** [comment]
See https://github.com/andlabs/libui/pull/20#issuecomment-211381971

**Comment by ereli** [comment]
I understand it shouldn't be the default state, but is there another way to pass `NSApp activateIgnoringOtherApps` or `NSApplicationActivateAllWindows` ?

**Comment by andlabs** [comment]
Recently I found that some features of OS X, namely Dictation, only work if the program is run in a .app bundle, which will remove the need for this. I need to do some more investigation to see if this is actually the right way to go...

**Comment by uchuugaka** [comment]
Have you considered trying one of:
`NSWorkspace.sharedWorkspace launchApplication: appName]`
https://developer.apple.com/documentation/appkit/nsworkspace/1531434-launchapplication?language=objc
`NSWorkspace.sharedWorkspace launchApplication: appName showIcon: showIconYesOrNo autolaunch: autolaunchYesOrNo]`
https://developer.apple.com/documentation/appkit/nsworkspace/1527334-launchapplication?language=objc
`[NSWorkspaces.sharedWorkspace launchApplicationAtURL: url options: opts configuration: config error: &error]`
https://developer.apple.com/documentation/appkit/nsworkspace/1534810-launchapplicationaturl?language=objc
`[NSWorkspaces.sharedWorkspace launchAppWithBundleIdentifier: bundleID options: opts additionalEventParamDescriptor: an_aedesc launchIdentifier: launchID]`
https://developer.apple.com/documentation/appkit/nsworkspace/1533335-launchappwithbundleidentifier?language=objc

There is no reason this would do anything other than what this activateIgnoringOtherApps does.
The options exist in NSWorkspace methods and were there in older deprecated LS (Launch Services) API functions. It is potentially rude and wrong to do it when opening a window but if that is at the launch of the application, then it is a normal process to activate, but not necessarily ignore all other windows.

The correct thing to do normally is to activate using normal policy at launch.
Most apps have a window presented at launch.
Active apps do not necessarily have windows.
They do normally have an icon and a main menu.

**Comment by andlabs** [comment]
Those methods are for starting a separate program from your own.

**Comment by andrewarrow** [comment]
looking for an ez solution for this one. Maybe after each build I can just throw the binary into the right folder structure and rename .app?

**Comment by uchuugaka** [comment]
> Those methods are for starting a separate program from within your own.

indeed! I must have been half asleep. Aaaaand, now I remember why I wrote that. The point was, those NSWorkspace methods are the equivalent of activation. You can absolutely tell the workspace to activate you in that way. Then, it in turn goes through Launch Services, because, those methods directly wrap the long deprecated LS prefixed Launch Services functions that could do the same. By going through Launch Services via NSWorkspace, it is behaving much much more the way a user activating you app in the UI behaves. It removes the unfair contention of focus stealing, removes the unnatural behavior and allows for user interactions on the system to order focus and activations as normal. It might seem backward and roundabout but it works.
